### DISCUSSION TITLE
Model3d set vertices and get mesh

### DISCUSSION THREAD

**Discussion by Snoopy137** [AlmasB/FXGL#1261] [Q&A]
Hi community,

I'm doing some experiment trying to animate a 3D model, I made some progress but the animation is still not so smooth, one of the problems is that the vertices need to be updated for the animation to happen but the 3D model has only vertices getter but no way to ser them, so you need to use a for to update one by one and that's just not optimum, I thought of editing the mesh vertices to have a better result but the method that returns the 3d meshes always gives me an empty list, is there any workaround to get the meshes or to set a a full list to the 3d model vertices?
Thanks in advance!

**Comment by AlmasB** [comment]
Hi, this can potentially be added to public API given 3D API is still very much in development. Can you provide some example code related to what should be possible through public API and I'll take a look. Thanks.

**Reply by Snoopy137** [reply]
Hey @AlmasB 

The code is herehttps://github.com/Snoopy137/FXGLthirdpersoncamera/tree/master

basically in that repository I'm trying to create a third person camera, sorry that is not to organize, I'm testing stuff, the intention is to then create classes using this functions to offere a third person camera configuration OOTB and a few other useful functionalities.
But about this issue, specific I try to use this method to populate the 3d model vertices but did not work
`model.getVertices().clear();
 model.getVertices().addAll(dukesAnimation.get(keyCycle.get()).getVertices());`

So I have to use a stream to set the x,y and z for each vertices, to change the position a I have a list of 3d models for each position of the animation, if you test the code you'll see that the first animation is kind of jittery, but then it start working pretty smooth, I just haven find the way to make it better yet, but I think having a simpler way to add the vertices would be the best approach.

For this project what I have is:
Third person camera features:
- Camera always looking at one entity and following it around.
- by changin where you call the player rotaiton camera can be set to determin the player direction or no.
- zoom in and out from the player.

Features being worked:
-aim function for third person shooter.
-camera fixed to player direction disabling movement (flight simulator generallyl have this kind of camera)
-aiming point
-3d gravity
-camera collision, right now if there's a wall or ground between duke and the camera you see the wall

**Reply by Snoopy137** [reply]
Hey @AlmasB 

Sorry for writing so much, I just thought of something, I think it would be useful to have a method to set vertices of a model3d, like the getVertices() but a setter, and also loading only vertices form an obj file, since the animation can be achieved by updating the vertices I don't think there's a need to load the entire object, just vertices and probably texture data

**Reply by AlmasB** [reply]
Ah I see, is there any particular format the animation is stored in? Based on the repo above, it looks like the animation is just a collection of .obj models, where each frame is a separate model.

At the moment, I'm not sure what the API should look like since the only valid use case for setting vertices (without setting faces or tex coords) is for an animation of this kind.

**Reply by Snoopy137** [reply]
Hi Almas, yes you are correct, for this case I made the animation in blender and when exported got the list of models, but loading a new model for each frame consumes a lot of resources, so I thought that instead I could simply update che vertex (I haven't tested this with texture yet). I loaded a list of models for testing, my idea would be to just load vertex and then set them to the model, currently for this I have to loop all vertex to ser x y and z points, which is better than loading an entire model, but the more vertices there are the more time it takes, so I thought having a way to set them all at once could improve the animation flow.
For the vertex loading I still have no ideas on how to implement that without loading the model at some point.
I think this could be a good approach for animating custom 3d models, but if I'm wrong you can just let me know and I'll keep investigating
Regards
